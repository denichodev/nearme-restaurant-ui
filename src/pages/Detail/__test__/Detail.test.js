import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import { Detail, Carousel } from '../index';

describe('<Detail />', () => {
  test('renders correctly', () => {
    const props = {
      restaurant: {
        loading: false,
        data: {
          additionalInfo: [],
          address:
            'Gandaria City, Lantai Upper Ground, Jl. Sultan Iskandar Muda, Gandaria, Jakarta',
          cover: 'http://localhost:8080/assets/top-momoiro.jpg',
          cuisine: 'Asian',
          id: 1,
          location: 'Kota Kasablanka, Tebet, Jakarta',
          name: 'Momoiro',
          openingHours: '10.00 - 22.00',
          phoneNumber: '08980780780',
          rating: 4.2,
          slides: [
            'http://localhost:8080/assets/momoiro-1.jpg',
            'http://localhost:8080/assets/momoiro-2.jpg',
            'http://localhost:8080/assets/momoiro-3.jpg',
          ],
          slug: 'momoiro-kota-kasablanka',
        },
        errors: [],
      },
      match: { params: { slug: '/' } },
      goBack: jest.fn(),
      getRestaurant: jest.fn(),
      reservation: {
        loading: false,
      },
      reservationToaster: {
        show: false,
        message: '',
      },
    };
    const wrapper = shallow(<Detail {...props} />);

    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  test('renders Carousel correctly', () => {
    const props = {
      className: 'gf-slide',
    };
    const wrapper = shallow(
      <Carousel {...props}>
        <div>
          <img src="random.jpg" />
        </div>
      </Carousel>,
    );

    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  test('renders loading correctly', () => {
    const props = {
      restaurant: {
        loading: true,
        data: {
          additionalInfo: [],
          address: '',
          cover: '',
          cuisine: '',
          id: 1,
          location: '',
          name: '',
          openingHours: '',
          phoneNumber: '',
          rating: 0,
          slides: [],
          slug: '',
        },
        errors: [],
      },
      match: { params: { slug: '/' } },
      goBack: jest.fn(),
      getRestaurant: jest.fn(),
      reservation: {
        loading: false,
      },
      reservationToaster: {
        show: false,
        message: '',
      },
    };
    const wrapper = shallow(<Detail {...props} />);

    expect(wrapper.find('.loading')).toHaveLength(1);
  });

  test('renders errors correctly', () => {
    const props = {
      restaurant: {
        loading: false,
        data: {
          additionalInfo: [],
          address: '',
          cover: '',
          cuisine: '',
          id: 1,
          location: '',
          name: '',
          openingHours: '',
          phoneNumber: '',
          rating: 0,
          slides: [],
          slug: '',
        },
        errors: ['Something went wrong'],
      },
      match: { params: { slug: '/' } },
      goBack: jest.fn(),
      getRestaurant: jest.fn(),
      reservation: {
        loading: false,
      },
      reservationToaster: {
        show: false,
        message: '',
      },
    };
    const wrapper = shallow(<Detail {...props} />);

    expect(wrapper.find('.errors')).toHaveLength(1);
  });

  test('renders overlay correctly', () => {
    const props = {
      restaurant: {
        loading: false,
        data: {
          additionalInfo: [],
          address: '',
          cover: '',
          cuisine: '',
          id: 1,
          location: '',
          name: '',
          openingHours: '',
          phoneNumber: '',
          rating: 0,
          slides: [],
          slug: '',
        },
        errors: [],
      },
      match: { params: { slug: '/' } },
      goBack: jest.fn(),
      getRestaurant: jest.fn(),
      reservation: {
        loading: false,
      },
      reservationToaster: {
        show: false,
        message: '',
      },
    };
    const wrapper = shallow(<Detail {...props} />);

    wrapper.find('.footerBookBtn button').simulate('click');
    expect(wrapper.find('.overlay')).toHaveLength(1);

    wrapper.find('.close').simulate('click');
    expect(wrapper.find('.overlay')).toHaveLength(0);
  });

  test('change state', () => {
    const props = {
      restaurant: {
        loading: false,
        data: {
          additionalInfo: [],
          address: '',
          cover: '',
          cuisine: '',
          id: 1,
          location: '',
          name: '',
          openingHours: '',
          phoneNumber: '',
          rating: 0,
          slides: [],
          slug: '',
        },
        errors: [],
      },
      match: { params: { slug: '/' } },
      goBack: jest.fn(),
      getRestaurant: jest.fn(),
      reservation: {
        loading: false,
      },
      reservationToaster: {
        show: false,
        message: '',
      },
    };
    const wrapper = shallow(<Detail {...props} />);

    wrapper.find('.footerBookBtn button').simulate('click');
    expect(wrapper.find('.overlay')).toHaveLength(1);

    wrapper.find('.close').simulate('click');
    expect(wrapper.find('.overlay')).toHaveLength(0);
  });

  test('renders button correctly', () => {
    const props = {
      restaurant: {
        loading: false,
        data: {
          additionalInfo: [],
          address: '',
          cover: '',
          cuisine: '',
          id: 1,
          location: '',
          name: '',
          openingHours: '',
          phoneNumber: '',
          rating: 0,
          slides: [],
          slug: '',
        },
        errors: [],
      },
      match: { params: { slug: '/' } },
      goBack: jest.fn(),
      getRestaurant: jest.fn(),
      reservation: {
        loading: true,
      },
      reservationToaster: {
        show: false,
        message: '',
      },
    };
    const wrapper = shallow(<Detail {...props} />);

    expect(wrapper.find('.reserve').text()).toEqual('Loading...');

    wrapper.setProps({ reservation: { loading: false } });
    expect(wrapper.find('.reserve').text()).toEqual('Make a reservation');
  });

  test('renders toaster correctly', () => {
    const props = {
      restaurant: {
        loading: false,
        data: {
          additionalInfo: [],
          address: '',
          cover: '',
          cuisine: '',
          id: 1,
          location: '',
          name: '',
          openingHours: '',
          phoneNumber: '',
          rating: 0,
          slides: [],
          slug: '',
        },
        errors: [],
      },
      match: { params: { slug: '/' } },
      goBack: jest.fn(),
      getRestaurant: jest.fn(),
      reservation: {
        loading: false,
      },
      reservationToaster: {
        show: true,
        message: 'Table reserved',
      },
    };
    const wrapper = shallow(<Detail {...props} />);

    expect(wrapper.find('.toaster')).toHaveLength(1);

    wrapper.setProps({ reservationToaster: { show: false, message: '' } });
    expect(wrapper.find('.toaster')).toHaveLength(0);
  });

  test('handle input change correctly', () => {
    const props = {
      restaurant: {
        loading: false,
        data: {
          additionalInfo: [],
          address: '',
          cover: '',
          cuisine: '',
          id: 1,
          location: '',
          name: '',
          openingHours: '',
          phoneNumber: '',
          rating: 0,
          slides: [],
          slug: '',
        },
        errors: [],
      },
      match: { params: { slug: '/' } },
      goBack: jest.fn(),
      getRestaurant: jest.fn(),
      reservation: {
        loading: false,
      },
      reservationToaster: {
        show: false,
        message: '',
      },
    };
    const wrapper = shallow(<Detail {...props} />);
    wrapper
      .find('.form input[type="number"]')
      .simulate('change', { target: { value: 2 } });
    expect(wrapper.state().form.guest).toEqual(2);
  });

  test('button should disabled if form is empty', () => {
    const props = {
      restaurant: {
        loading: false,
        data: {
          additionalInfo: [],
          address: '',
          cover: '',
          cuisine: '',
          id: 1,
          location: '',
          name: '',
          openingHours: '',
          phoneNumber: '',
          rating: 0,
          slides: [],
          slug: '',
        },
        errors: [],
      },
      match: { params: { slug: '/' } },
      goBack: jest.fn(),
      getRestaurant: jest.fn(),
      reservation: {
        loading: false,
      },
      reservationToaster: {
        show: false,
        message: '',
      },
    };
    const wrapper = shallow(<Detail {...props} />);
    wrapper.setState({ form: { checkIn: '', guest: 0 } });
    expect(wrapper.find('.form button').prop('disabled')).toEqual(true);
  });

  test('submit button should reset form', () => {
    const props = {
      restaurant: {
        loading: false,
        data: {
          additionalInfo: [],
          address: '',
          cover: '',
          cuisine: '',
          id: 1,
          location: '',
          name: '',
          openingHours: '',
          phoneNumber: '',
          rating: 0,
          slides: [],
          slug: '',
        },
        errors: [],
      },
      match: { params: { slug: '/' } },
      goBack: jest.fn(),
      getRestaurant: jest.fn(),
      reservation: {
        loading: false,
      },
      reservationToaster: {
        show: false,
        message: '',
      },
      submitReservation: jest.fn(),
    };
    const wrapper = shallow(<Detail {...props} />);

    wrapper.setState({ form: { checkIn: '2018:02:20 00:00:00', guest: 2 } });
    expect(wrapper.state().form).toEqual({
      checkIn: '2018:02:20 00:00:00',
      guest: 2,
    });

    wrapper.find('.form button').simulate('click', { preventDefault() {} });
    expect(wrapper.state().form).toEqual({
      checkIn: '',
      guest: 0,
    });

    expect(props.submitReservation).toBeCalled();
  });

  test('input should call handleChange', () => {
    const props = {
      restaurant: {
        loading: false,
        data: {
          additionalInfo: [],
          address: '',
          cover: '',
          cuisine: '',
          id: 1,
          location: '',
          name: '',
          openingHours: '',
          phoneNumber: '',
          rating: 0,
          slides: [],
          slug: '',
        },
        errors: [],
      },
      match: { params: { slug: '/' } },
      goBack: jest.fn(),
      getRestaurant: jest.fn(),
      reservation: {
        loading: false,
      },
      reservationToaster: {
        show: false,
        message: '',
      },
      submitReservation: jest.fn(),
    };
    const wrapper = shallow(<Detail {...props} />);
    const validateForm = jest.spyOn(wrapper.instance(), 'validateForm');
    const event = { target: { value: 2 } };

    wrapper.find('.form .guest').simulate('change', event);

    expect(validateForm).toHaveBeenCalled();
    expect(wrapper.state().form).toEqual({ checkIn: null, guest: 2 });
  });

  test('hideToaster called after 3s', () => {
    const props = {
      restaurant: {
        loading: false,
        data: {
          additionalInfo: [],
          address: '',
          cover: '',
          cuisine: '',
          id: 1,
          location: '',
          name: '',
          openingHours: '',
          phoneNumber: '',
          rating: 0,
          slides: [],
          slug: '',
        },
        errors: [],
      },
      match: { params: { slug: '/' } },
      goBack: jest.fn(),
      getRestaurant: jest.fn(),
      reservation: {
        loading: false,
      },
      reservationToaster: {
        show: false,
        message: '',
      },
      submitReservation: jest.fn(),
    };
    const wrapper = shallow(<Detail {...props} />);

    wrapper.setProps({
      reservationToaster: { show: true, message: 'Table reserved' },
    });

    expect(wrapper.find('.toaster')).toHaveLength(1);
  });
});
