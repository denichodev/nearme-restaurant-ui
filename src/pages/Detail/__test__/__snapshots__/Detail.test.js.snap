// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Detail /> renders Carousel correctly 1`] = `
<Slider
  arrows={false}
  className="gf-slide"
  infinite={false}
  lazyLoad={true}
  responsive={
    Array [
      Object {
        "breakpoint": 600,
        "centerPadding": "0px",
        "settings": Object {
          "slidesToShow": 1,
        },
      },
      Object {
        "breakpoint": 1920,
        "centerPadding": "0px",
        "settings": Object {
          "slidesToShow": 2,
        },
      },
    ]
  }
>
  <div>
    <img
      src="random.jpg"
    />
  </div>
</Slider>
`;

exports[`<Detail /> renders correctly 1`] = `
<div>
  <HelmetWrapper
    defer={true}
    encodeSpecialCharacters={true}
  >
    <body
      className="detail-page"
    />
    <link
      charset="UTF-8"
      href="https://cdnjs.cloudflare.com/ajax/libs/slick-carousel/1.6.0/slick.min.css"
      rel="stylesheet"
      type="text/css"
    />
    <link
      href="https://cdnjs.cloudflare.com/ajax/libs/slick-carousel/1.6.0/slick-theme.min.css"
      rel="stylesheet"
      type="text/css"
    />
  </HelmetWrapper>
  <div
    className="mobileHeader"
  >
    <div
      className="backButton"
    >
      <button
        onClick={[MockFunction]}
      >
        <i
          className="fas fa-angle-left"
        />
      </button>
    </div>
  </div>
  <Component
    className="slickSlider"
  >
    <div
      key="0"
    >
      <img
        alt="test"
        src="http://localhost:8080/assets/momoiro-1.jpg"
      />
    </div>
    <div
      key="1"
    >
      <img
        alt="test"
        src="http://localhost:8080/assets/momoiro-2.jpg"
      />
    </div>
    <div
      key="2"
    >
      <img
        alt="test"
        src="http://localhost:8080/assets/momoiro-3.jpg"
      />
    </div>
  </Component>
  <Grid
    className="detailPage"
    fluid={false}
  >
    <Row>
      <Col
        md={8}
        sm={12}
        xs={12}
      >
        <div
          className="wrapper"
        >
          <Main
            additionalInfo={Array []}
            cuisine="Asian"
            name="Momoiro"
            rating={4.2}
          />
          <Info
            address="Gandaria City, Lantai Upper Ground, Jl. Sultan Iskandar Muda, Gandaria, Jakarta"
            cuisine="Asian"
            openingHours="10.00 - 22.00"
            phoneNumber="08980780780"
          />
        </div>
      </Col>
      <Col
        md={4}
        sm={12}
        xs={12}
      >
        <aside>
          <div
            className="sidebar"
          >
            <button
              className="close"
              onClick={[Function]}
            >
              <i
                className="fas fa-times"
              />
            </button>
            <div
              className="form"
            >
              <Component
                className=""
                closeOnSelect={false}
                closeOnTab={true}
                dateFormat={true}
                defaultValue=""
                input={false}
                inputProps={
                  Object {
                    "placeholder": "Select date and time...",
                  }
                }
                isValidDate={[Function]}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                onViewModeChange={[Function]}
                strictParsing={true}
                timeConstraints={Object {}}
                timeFormat={true}
                utc={false}
                value={null}
              />
            </div>
            <div
              className="form"
            >
              <input
                className="guest"
                onChange={[Function]}
                placeholder="Total guest..."
                type="number"
                value={0}
              />
              <div
                className="person"
              >
                 person
              </div>
            </div>
            <div
              className="form"
            >
              <button
                className="reserve"
                disabled={true}
                onClick={[Function]}
              >
                Make a reservation
              </button>
            </div>
          </div>
        </aside>
        <div
          className="footerBookBtn"
        >
          <button
            onClick={[Function]}
          >
            Make a reservation
          </button>
        </div>
      </Col>
    </Row>
  </Grid>
</div>
`;
